reactfire/firebaseApp/sdk.tsx

```diff
@@ -1,218 +1,117 @@
-  firebaseApp?: firebase.app.App,
-  settingsCallback?: (functions: () => Functions) => void
-) {
-  return fetchSDK(SDK.FUNCTIONS, firebaseApp, settingsCallback);
-}
-
-export function useFunctions(firebaseApp?: firebase.app.App) {
-  return useSDK(SDK.FUNCTIONS, firebaseApp);
-}
-
-export function preloadMessaging(
-  firebaseApp: firebase.app.App,
-  settingsCallback?: (messaging: () => Messaging) => void
-) {
-  return fetchSDK(SDK.MESSAGING, firebaseApp, settingsCallback);
-}
-
-export function useMessaging(firebaseApp?: firebase.app.App) {
-  return useSDK(SDK.MESSAGING, firebaseApp);
-}
+let firebase: typeof import('firebase/app') | undefined;
 
-export function preloadPerformance(
-  firebaseApp: firebase.app.App,
-  settingsCallback?: (performance: () => Performance) => void
+function useFirebaseComponent(
+  componentName: ComponentName
 ) {
-  return fetchSDK(SDK.PERFORMANCE, firebaseApp, settingsCallback);
-}
-
-export function usePerformance(firebaseApp?: firebase.app.App) {
-  return useSDK(SDK.PERFORMANCE, firebaseApp);
-}
-
-export function preloadRemoteConfig(
-  firebaseApp: firebase.app.App,
-  settingsCallback?: (remoteConfig: () => RemoteConfig) => Promise<any>
-) {
-  return fetchSDK(SDK.REMOTE_CONFIG, firebaseApp, settingsCallback);
-}
-
-export function useRemoteConfig(firebaseApp?: firebase.app.App) {
-  return useSDK(
-    SDK.REMOTE_CONFIG,
-    firebaseApp
-  ) as () => firebase.remoteConfig.RemoteConfig;
-}
-
-export function preloadStorage(
-  firebaseApp: firebase.app.App,
-  settingsCallback: (storage: () => Storage) => Promise<void>
-) {
-  return fetchSDK(SDK.STORAGE, firebaseApp, settingsCallback);
-}
-
-export function useStorage(firebaseApp?: firebase.app.App) {
-  return useSDK(SDK.STORAGE, firebaseApp);
-}
+  if (!firebase) { throw import('firebase/app').then(it => firebase = it); }
+  if (!firebase[componentName]) { throw importSDK(componentName); }
+  return firebase[componentName];
+}
+
+function proxyComponent(componentName: "auth"        ): typeof import('firebase/app').auth;
+function proxyComponent(componentName: "analytics"   ): typeof import('firebase/app').analytics;
+function proxyComponent(componentName: "database"    ): typeof import('firebase/app').database;
+function proxyComponent(componentName: "firestore"   ): typeof import('firebase/app').firestore;
+function proxyComponent(componentName: "functions"   ): typeof import('firebase/app').functions;
+function proxyComponent(componentName: "messaging"   ): typeof import('firebase/app').messaging;
+function proxyComponent(componentName: "performance" ): typeof import('firebase/app').performance;
+function proxyComponent(componentName: "remoteConfig"): typeof import('firebase/app').remoteConfig;
+function proxyComponent(componentName: "storage"     ): typeof import('firebase/app').storage;
+function proxyComponent(componentName: ComponentName): FirebaseNamespaceComponent {
+  return new Proxy(() => useFirebaseComponent(componentName), {
```

Comment:
Downside, I can't seem to use react context within the proxy... keep investigating else I won't be able to drop the `()()` style.